File: Sources/OpenDDSharp/TransportInstManager.cpp
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 - 2021 Jose Morato

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenDDSharp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with OpenDDSharp. If not, see <http://www.gnu.org/licenses/>.
**********************************************************************/
#include "TransportInstManager.h"

OpenDDSharp::TransportInstManager* OpenDDSharp::TransportInstManager::get_instance() {
    return ACE_Singleton<TransportInstManager, ACE_SYNCH_MUTEX>::instance();
}

void OpenDDSharp::TransportInstManager::add(::OpenDDS::DCPS::TransportInst* native, gcroot<OpenDDSharp::OpenDDS::DCPS::TransportInst^> managed) {
    m_insts.emplace(native, managed);
}

void OpenDDSharp::TransportInstManager::remove(::OpenDDS::DCPS::TransportInst* native) {
    m_insts.erase(native);
}

gcroot<OpenDDSharp::OpenDDS::DCPS::TransportInst^> OpenDDSharp::TransportInstManager::find(::OpenDDS::DCPS::TransportInst* native) {
    std::map<::OpenDDS::DCPS::TransportInst*, gcroot<OpenDDSharp::OpenDDS::DCPS::TransportInst^>>::iterator it;
    it = m_insts.find(native);
    if (it != m_insts.end()) {
        return it->second;
    }
    else {
        return nullptr;
    }
}

File: Sources/OpenDDSharp/OpenDDS/DCPS/TransportInstManager.cs
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 Jose Morato - OpenDDSharp

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the MIT License.
**********************************************************************/
using System;
using System.Collections.Concurrent;

namespace OpenDDSharp.OpenDDS.DCPS;

internal class TransportInstManager
{
    #region Fields
    private static readonly object _lock = new object();
    private static TransportInstManager _instance;
    private readonly ConcurrentDictionary<IntPtr, TransportInst> _insts = new ();
    #endregion

    #region Singleton
    public static TransportInstManager Instance
    {
        get
        {
            lock (_lock)
            {
                return _instance ??= new TransportInstManager();
            }
        }
    }
    #endregion

    #region Methods
    public void Add(IntPtr ptr, TransportInst inst)
    {
        _insts.AddOrUpdate(ptr, inst, (p, t) => inst);
    }

    public void Remove(IntPtr ptr)
    {
        _insts.TryRemove(ptr, out _);
    }

    public TransportInst Find(IntPtr ptr)
    {
        return _insts.TryGetValue(ptr, out var found) ? found : null;
    }

    public void Clear()
    {
        _insts.Clear();
    }
    #endregion
}